src/app.service.ts

```diff
@@ -1,8 +1,158 @@
 import { Injectable } from '@nestjs/common';
+import axios from 'axios';
 
 @Injectable()
 export class AppService {
+  constructor(private dataSource: any) {}
   getHello(): string {
     return 'Hello World!';
   }
+
+  calculate(array: number[]): { max: number; sum: number } {
+    let max = -Infinity;
+    let sum = 0;
+
+    for (let i = 0; i < array.length; i++) {
+      if (array[i] > max) {
+        max = array[i]; // Find the maximum
+      }
+      sum += array[i]; // Calculate the sum
+    }
+
+    return { max, sum };
+  }
+
+  getGreetingVer2(name: string, timeOfDay: string): string {
+    if (timeOfDay === 'morning') {
+      return `Good morning, ${name}`;
+    } else if (timeOfDay === 'afternoon') {
+      return `Good afternoon, ${name}`;
+    } else if (timeOfDay === 'evening') {
+      return `Good evening, ${name}`;
+    } else {
+      return `Hello, ${name}`;
+    }
+  }
+
+  isPalindrome(str: string): boolean {
+    str = str.toLowerCase().replace(/[^a-z0-9]/g, ''); // Remove non-alphanumeric characters
+    let reversed = '';
+
+    // Inefficient: Manually reverse the string
+    for (let i = str.length - 1; i >= 0; i--) {
+      reversed += str[i]; // String concatenation in a loop
+    }
+
+    return str === reversed;
+  }
+
+  async getTotalSalesForYear(year: number): Promise<number> {
+    const query = `
+      SELECT SUM(TotalAmount)
+      FROM Orders
+      WHERE EXTRACT(YEAR FROM OrderDate) = $1
+    `;
+
+    const result = await this.dataSource.query(query, [year]);
+    return result[0]?.sum || 0;
+  }
+
+  async getCustomerOrders(customerId: number): Promise<any[]> {
+    const query = `
+      SELECT *
+      FROM Orders
+      WHERE CustomerID = $1
+      ORDER BY DATE(OrderDate)
+    `;
+
+    return this.dataSource.query(query, [customerId]);
+  }
+
+  async getUserByUsername(username: string): Promise<any> {
+    // Vulnerable to SQL injection
+    const query = `SELECT * FROM Users WHERE username = '${username}'`;
+    return this.dataSource.query(query);
```

Comment:
<div>


<div id="suggestion">
<div id="issue"><b>SQL injection vulnerability in getUserByUsername</b></div>
<div id="fix">

Consider using parameterized queries with `?` placeholders instead of string interpolation to prevent SQL injection vulnerabilities. The current implementation directly interpolates user input into the SQL query.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<blockquote>Check the AI-generated fix before applying</blockquote>
<div id="code">


```
 @@ -71,4 +71,4 @@
  async getUserByUsername(username: string): Promise<any> {
      // Vulnerable to SQL injection
 -    const query = `SELECT * FROM Users WHERE username = '${username}'`;
 +    const query = `SELECT * FROM Users WHERE username = ?`;
      return this.dataSource.query(query, [username]);
```

</div>
</details>
</div>


<small><i>Code Review Run <a href=https://github.com/phtrung212/test-app/pull/35#issuecomment-2561677458>#95560c</a></i></small>
</div>

---
Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged